evm/src/cross_table_lookup.rs

```diff
@@ -143,13 +188,25 @@ pub fn cross_table_lookup_data<F: RichField, C: GenericConfig<D, F = F>, const D
 fn partial_products<F: Field>(
     trace: &[PolynomialValues<F>],
     columns: &[usize],
+    filter_columns: &[usize],
     challenge: GrandProductChallenge<F>,
 ) -> PolynomialValues<F> {
     let mut partial_prod = F::ONE;
     let degree = trace[0].len();
     let mut res = Vec::with_capacity(degree);
     for i in 0..degree {
-        partial_prod *= challenge.combine(columns.iter().map(|&j| &trace[j].values[i]));
+        let filter = if filter_columns.is_empty() {
+            F::ONE
+        } else {
+            filter_columns.iter().map(|&j| trace[j].values[i]).sum()
+        };
+        partial_prod *= if filter.is_zero() {
+            F::ONE
+        } else if filter.is_one() {
+            challenge.combine(columns.iter().map(|&j| &trace[j].values[i]))
+        } else {
+            panic!("Non-binary filter?")
+        };
```

Comment:
Since this might be slightly performance sensitive, might be best to avoid multiplication by 1 like
```suggestion
        if filter.is_one() {
            partial_prod *= challenge.combine(columns.iter().map(|&j| &trace[j].values[i]));
        } else {
            assert_eq!(filter, F::ZERO, "Non-binary filter?");
        }
```